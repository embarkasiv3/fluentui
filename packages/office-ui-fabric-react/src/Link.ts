export * from './packages/react-fundamentals/Link';
