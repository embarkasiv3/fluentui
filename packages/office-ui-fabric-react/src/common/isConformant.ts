import { isConformant as isConformantBase, IsConformantOptions } from '@fluentui/react-conformance';

export function isConformant(testInfo: IsConformantOptions) {
  const defaultOptions = {
    disabledTests: ['has-docblock'],
  };

  isConformantBase(defaultOptions, testInfo);
}
