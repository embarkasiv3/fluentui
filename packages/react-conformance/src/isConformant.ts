import { TestObject, TestingOptions } from './types';
import { withCustomConfig } from 'react-docgen-typescript';
import * as path from 'path';
import { defaultTests } from './defaultTests';

export function isConformant(testInfo: TestingOptions, extraTests?: TestObject) {
  const { componentPath, displayName } = testInfo;
  const tsconfigPath = path.join(process.cwd(), 'tsconfig.json');
  const parser = withCustomConfig(tsconfigPath, {});
  const components = parser.parse(componentPath);
  const mainComponents = components.filter(comp => comp.displayName === displayName);

  if (mainComponents.length === 1) {
    const componentInfo = mainComponents[0];

    for (const test of Object.keys(defaultTests)) {
      defaultTests[test](componentInfo, testInfo);
    }
    if (extraTests) {
      for (const test of Object.keys(extraTests)) {
        extraTests[test](componentInfo, testInfo);
      }
    }
  } else {
    if (components.length === 0) {
      throw new Error('No exported components in path: ' + componentPath);
    } else {
      throw new Error(
        `No component with name '${displayName}' was found at ${componentPath}. ` +
          `These are the exported component names: ${components.map(component => component.displayName).join(', ')}`,
      );
    }
  }
}
