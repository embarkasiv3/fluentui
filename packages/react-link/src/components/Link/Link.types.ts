import * as React from 'react';
import { ComponentProps } from '@fluentui/react-utilities';

/**
 * {@docCategory Link}
 */
export type LinkProps = ComponentProps &
  React.AnchorHTMLAttributes<HTMLAnchorElement | HTMLButtonElement | HTMLElement> &
  Omit<React.ButtonHTMLAttributes<HTMLAnchorElement | HTMLButtonElement | HTMLElement>, 'type'> & {
    /**
     * Whether the link is disabled.
     * @defaultvalue false
     */
    disabled?: boolean;

    /**
     * When set, allows the link to be focusable even when it has been disabled. This is used in scenarios where it is
     * important to keep a consistent tab order for screen reader and keyboard users.
     * @defaultvalue false
     */
    disabledFocusable?: boolean;

    /**
     * URL the link points to. If not provided, the link renders as a button (unless that behavior is
     * overridden using `as`).
     */
    href?: string;

    /**
     * If true, changes styling when the link is being used alongside other text content.
     */
    inline?: boolean;

    /**
     * Click handler for the link.
     */
    onClick?: (event: React.MouseEvent<HTMLAnchorElement | HTMLButtonElement | HTMLElement>) => void;

    /**
     * If true, changes the link styling to emphasize that it represents an alternative action.
     * @defaultvalue false
     */
    secondary?: boolean;

    /**
     * Relationship to the linked URL (can be a space-separated list).
     * Most common values are `noreferrer` and/or `noopener`.
     * This prop is only applied if `href` is set.
     */
    rel?: string;

    /**
     * Where to open the linked URL. Common values are `_blank` (a new tab or window),
     * `_self` (the current window/context), `_parent`, and `_top`.
     * This prop is only applied if `href` is set.
     */
    target?: string;

    /**
     * Built-in HTML attribute with different behavior depending on how the link is rendered.
     * If rendered as `<a>`, hints at the MIME type.
     * If rendered as `<button>`, override the type of button (`button` is the default).
     */
    type?: string;

    /** Style tokens */
    // tokens?: RecursivePartial<LinkTokens>;
  };

/**
 * {@docCategory Link}
 */
export interface LinkState extends LinkProps {}

// /**
//  * {@docCategory Link}
//  */
// export type LinkTokens = ColorTokens &
//   FontTokens & {
//     // The text decoration used for the link in its default state.
//     textDecoration?: string;

//     // The set of color tokens that are applied when the link has been visited.
//     visited?: ColorTokenSet;

//     // The text decoration used for the link when it is being focused.
//     focused?: {
//       textDecoration: string;
//     };

//     // The text decoration used for the link when it is being hovered.
//     hovered?: {
//       textDecoration: string;
//     };

//     // The text decoration used for the link when it is being pressed.
//     pressed?: {
//       textDecoration: string;
//     };

//     // The text decoration used for the link when it is in a disabled state.
//     disabled?: {
//       textDecoration: string;
//     };
//   };

// /**
//  * {@docCategory Link}
//  */
// export type LinkVariants<TTokens = LinkTokens> = {
//   root?: TTokens;
// };
