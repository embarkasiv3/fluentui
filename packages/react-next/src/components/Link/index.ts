export * from './Link';
export * from './LinkBase';
export * from './Link.types';
export * from './useLink';
