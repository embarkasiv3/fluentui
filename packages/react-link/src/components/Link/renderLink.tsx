import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import { LinkState } from './Link.types';
import { linkShorthandProps } from './useLink';

/**
 * Defines the render function. Given the state of a Link, renders it.
 */
export const renderLink = (state: LinkState) => {
  const { slots, slotProps } = getSlots(state, linkShorthandProps);

  return <slots.root {...slotProps.root} />;
};
