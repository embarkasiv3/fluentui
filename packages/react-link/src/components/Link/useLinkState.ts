import * as React from 'react';
import { getCode, EnterKey, SpacebarKey } from '@fluentui/keyboard-key';
import { LinkState } from './Link.types';

/**
 * The useLink hook processes the Link draft state.
 * @param state - Link draft state to mutate.
 */
export const useLinkState = (state: LinkState) => {
  const { as, disabled, disabledFocusable, href, onClick, onKeyDown: onKeyDownCallback, type } = state;

  state.onClick = (ev: React.MouseEvent<HTMLAnchorElement | HTMLButtonElement | HTMLElement>) => {
    if (disabled) {
      ev.preventDefault();
    } else {
      onClick?.(ev);
    }
  };

  const onKeyDown = (ev: React.KeyboardEvent<HTMLAnchorElement | HTMLButtonElement | HTMLElement>) => {
    const keyCode = getCode(ev);
    const isEnterOrSpaceKey = keyCode === EnterKey || keyCode === SpacebarKey;
    if (disabled && isEnterOrSpaceKey) {
      ev.preventDefault();
      ev.stopPropagation();
    } else {
      onKeyDownCallback?.(ev);

      if (onClick && isEnterOrSpaceKey) {
        onClick((ev as unknown) as React.MouseEvent<HTMLAnchorElement | HTMLButtonElement | HTMLElement>);
      }
    }
  };

  // Adjust props depending on the root type.
  if (typeof as === 'string') {
    if (as === 'a') {
      state.href = disabled ? undefined : href;
      state.tabIndex = disabled && !disabledFocusable ? undefined : 0;
    } else {
      // Remove the href, rel and target props for all non-anchor elements.
      state.href = undefined;
      state.rel = undefined;
      state.target = undefined;

      // Add the type="button" prop for button elements
      if (as === 'button') {
        state.type = type ? type : 'button';
      }
      // Add keydown event handler for all other elements
      else {
        state.onKeyDown = onKeyDown;
      }
    }
  }
  // Add keydown event handler for all other elements
  else {
    state.onKeyDown = onKeyDown;
  }

  // Add aria attributes
  state['aria-disabled'] = disabled || disabledFocusable;
  state.disabled = as === 'a' || as === 'button' ? disabled && !disabledFocusable : undefined;
};
