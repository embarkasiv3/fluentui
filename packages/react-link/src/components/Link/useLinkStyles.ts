import { ax, makeStyles } from '@fluentui/react-make-styles';
import { LinkState } from './Link.types';

const useStyles = makeStyles<LinkState>([
  // Common styles.
  [
    null,
    tokens => ({
      backgroundColor: 'transparent',
      border: 'none',
      borderBottom: 'solid transparent',
      borderBottomWidth: tokens.global.strokeWidth.thin,
      boxSizing: 'border-box',
      color: tokens.global.palette.brand.primary,
      cursor: 'pointer',
      display: 'inline',
      fontFamily: tokens.global.type.fontFamilies.base,
      fontSize: tokens.global.type.fontSizes.base[300],
      fontWeight: tokens.global.type.fontWeights.regular,
      margin: 0,
      overflow: 'inherit',
      padding: 0,
      textAlign: tokens.global.type.alignment.start,
      textDecoration: 'none',
      textOverflow: 'inherit',
      userSelect: 'text',

      ':hover': {
        borderBottomColor: tokens.global.palette.brand.primary,
      },

      ':active': {
        borderBottomColor: tokens.global.palette.brand.primary,
      },
    }),
  ],

  // Overrides when an href is present so the Link renders as an anchor.
  [
    s => !!s.href,
    {
      fontSize: 'inherit',
    },
  ],

  // Overrides when the Link is rendered inline within text.
  [
    s => s.inline,
    tokens => ({
      borderBottomColor: tokens.global.palette.brand.primary,
    }),
  ],

  // Overrides when the Link is disabled.
  [
    s => s['aria-disabled'] as boolean | undefined,
    tokens => ({
      borderBottomColor: 'transparent',
      color: tokens.global.palette.grey[74],
      cursor: 'not-allowed',

      ':hover': {
        borderBottomColor: 'transparent',
      },

      ':active': {
        borderBottomColor: 'transparent',
      },
    }),
  ],
]);

export const useLinkStyles = (state: LinkState): LinkState => {
  state.className = ax(useStyles(state), state.className);

  return state;
};
