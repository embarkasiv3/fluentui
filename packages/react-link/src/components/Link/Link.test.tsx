import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { ReactWrapper } from 'enzyme';
import { isConformant } from '../../common/isConformant';
import { Link } from './Link';

describe('Link', () => {
  isConformant({
    Component: Link,
    displayName: 'Link',
  });

  let wrapper: ReactWrapper | undefined;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  /**
   * Note: see more visual regression tests for Link in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const component = renderer.create(<Link>Default Link</Link>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
