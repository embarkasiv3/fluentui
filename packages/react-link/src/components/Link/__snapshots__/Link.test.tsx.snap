// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders a default state 1`] = `
<button
  className=""
  onClick={[Function]}
  type="button"
>
  Default Link
</button>
`;
